mmaction/datasets/pipelines/augmentations.py

```diff
@@ -868,6 +870,406 @@ def __repr__(self):
+            trial = 0
+            sample = 0
+            max_trials = cuboid_setting['max_trials']
+            max_sample = cuboid_setting['max_sample']
+            sampler = cuboid_setting['sampler']
+
+            min_scale = sampler.get('min_scale', 1)
+            max_scale = sampler.get('max_scale', 1)
+            min_aspect = sampler.get('min_aspect', 1)
+            max_aspect = sampler.get('max_aspect', 1)
+
+            while trial < max_trials and sample < max_sample:
+                # sample a normalized box
+                scale = np.random.uniform(min_scale, max_scale)
+                aspect = np.random.uniform(min_aspect, max_aspect)
+
+                w = scale * np.sqrt(aspect)
+                h = scale / np.sqrt(aspect)
+                if w > 1 or h > 1:
+                    continue
+                x = np.random.uniform(0, 1 - w)
+                y = np.random.uniform(0, 1 - h)
+
+                # rescale the box
+                sampled_cuboid = np.array(
+                    [x * img_w, y * img_h, (x + w) * img_w, (y + h) * img_h],
+                    dtype=np.float32)
+                # check constraint
+                trial += 1
+                if 'constraints' not in cuboid_setting:
+                    sampled_cuboids.append(sampled_cuboid)
+                    sample += 1
+                    continue
+
+                contraints = cuboid_setting['constraints']
+                min_jaccard_overlap = contraints.get('min_jaccard_overlap',
+                                                     np.inf)
+                max_jaccard_overlap = contraints.get('max_jaccard_overlap',
+                                                     np.inf)
+                ious = np.array(
+                    [np.mean(iou2d(tube, sampled_cuboid)) for tube in tubes])
+
+                if ious.size == 0:
+                    # empty gt
+                    sample += 1
+                    continue
+
+                if ious.max() >= min_jaccard_overlap or ious.min(
+                ) >= max_jaccard_overlap:
+                    sampled_cuboids.append(sampled_cuboid)
+                    sample += 1
+                    continue
+
+        return sampled_cuboids
+
+    def __call__(self, results):
+        img_h, img_w = results['img_shape']
+        gt_bboxes = results['gt_bboxes']
+        imgs = results['imgs']
+
+        candidate_cuboids = self.sample_cuboids(gt_bboxes, img_h, img_w)
+        if len(candidate_cuboids) == 0:
+            return results
+
+        cuboid = random.choice(candidate_cuboids)
+        x1, y1, x2, y2 = map(int, cuboid.tolist())
+
+        for i in range(len(imgs)):
+            imgs[i] = imgs[i][y1:y2, x1:x2, :]
+        out_tubes = defaultdict(list)
+        width = x2 - x1
+        height = y2 - y1
+
+        for label_index in gt_bboxes:
+            for tube in gt_bboxes[label_index]:
+                tube -= np.array([[x1, y1, x1, y1]], dtype=np.float32)
```

Comment:
double check you want to make inplace changes to gt_bboxes

Reply:
![image](https://user-images.githubusercontent.com/33949370/100619748-8bc4c800-3358-11eb-8070-67f8f08366e8.png)